### ISSUE TITLE
Limit typeparams in Generic Fake creation methods to reference types

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#1465]
I don't think this has been discussed before, but maybe I'm forgetting.

The `A.Fake` method could be defined as 

```c#
public static T Fake<T>() where T : class
```

(As could all other generic Fake creation methods.)

That way users would be prevented _by the compiler_ from trying to fake a struct or similar.

The helpful error message we output when they _do_ try doesn't really provide a lot more information than the typeparam constraint would:

```
Failed to create fake of type FakeItEasy.Specs.CreationSpecsBase+Struct:
  The type of proxy must be an interface or a class but it was FakeItEasy.Specs.CreationSpecsBase+Struct.
```

so I'm not sure we'd be losing much, if anything.

**Comment by thomaslevesque** [comment]
Sounds like a good idea. `A.Dummy`, on the other hand, would remain unconstrained.
There might be other places in the API where a constraint would be useful too.
BTW, it would be a breaking change

**Comment by thomaslevesque** [comment]
This change has been released in FakeItEasy [5.0.0](https://github.com/FakeItEasy/FakeItEasy/releases/5.0.0).
